src/AndroidDebugLauncher/Launcher.cs

```diff
@@ -556,37 +556,63 @@ private void VerifySdkVersion()
             string gdbServerPath = workingDirectory + "/lib/gdbserver";
 
             string lsCommand = string.Format(CultureInfo.InvariantCulture, "ls {0}", gdbServerPath);
             string output = ExecCommand(lsCommand);
-            if (string.Compare(output, gdbServerPath, StringComparison.OrdinalIgnoreCase) == 0)
-            {
-                //gdbserver is symlinked correctly
-                return gdbServerPath;
-            }
-            else if (_launchOptions.TargetArchitecture == TargetArchitecture.X64)
-            {
-                //start looking other places, only do this on 64 bit
-                //TODO: This needs some additional testsing
-                lsCommand = string.Format(CultureInfo.InvariantCulture, "ls /data/app/{0}*/lib/x86_64/gdbserver", _launchOptions.Package);
-                output = ExecCommand(lsCommand);
-                return output;
-            }
-            else
-            {
-                throw new LauncherException(Telemetry.LaunchFailureCode.NoGdbServer, string.Format(CultureInfo.CurrentCulture, LauncherResources.Error_NoGdbServer, gdbServerPath));
+            if (string.Compare(output, gdbServerPath, StringComparison.OrdinalIgnoreCase) != 0)
+            {
+                // <app data>/lib is not symlinked to system's app data directory (/data/app/<app package>/lib). 
+                // We copy gdbserver ourselves because of 2 reasons:
+                // 1. the app user doesn't have permissions on /data/app/<app package>-<suffix>/lib (it's owned by system)
+                // 2. we can't deterministically figure out the location where gdbserver is in /data/app because of the suffix,
+                //    which is used by Android to have multiple copies of the APK extracted in the system.
+                string arch = _launchOptions.TargetArchitecture.ToNDKPath();
+                string tmpGDBServer = string.Format(CultureInfo.InvariantCulture, "/data/local/tmp/{0}-gdbserver", arch);
+
+                device.FileSystem.Upload(_installPaths.GDBServerPath, tmpGDBServer, true);
+
+                // Files can't be executed from /data/local/tmp
+                string dest = string.Format(CultureInfo.InvariantCulture, "{0}/{1}-gdbserver", workingDirectory, arch);
+
+                // Can't use cp since shell doesn't have permissions to /data/data/<app>
+                // and the app uid doesn't have permissions to /data/local/tmp
+                string copyGDBServerCommand = string.Format(CultureInfo.InvariantCulture,
+                    "cat {0} | run-as {1} /system/bin/sh -c \"cat > {2}\"", tmpGDBServer, _launchOptions.Package, dest);
+                int rc = ExecCommand(copyGDBServerCommand, out output);
+                RunAsOutputParser.ThrowIfRunAsErrors(output, _launchOptions.Package);
+                if (rc != 0)
+                {
+                    throw new LauncherException(
+                        Telemetry.LaunchFailureCode.NoGdbServer,
```

Comment:
Here and in line 608 I used the `NoGdbServer` telemetry event since we only go down this code path if gdbserver can't be found. Let me know if I should create a different telemetry event.